chai/source.py

```diff
@@ -0,0 +1,169 @@
+                in_comma_sep_deps = True
+            else:
+                new_deps = _get_dep_from_instance_line(line)
+                deps.extend(new_deps)
+
+        if in_comma_sep_deps:
+            new_deps = _get_comma_separated_deps(line)
+            # Check if we have a blank line,
+            # in which case we don't add anything and keep searching
+            if new_deps:
+                deps.extend(new_deps)
+                # If the line ends in a comma, we will have more deps to come
+                if not line.rstrip().endswith(","):
+                    in_comma_sep_deps = False
+
+    return deps
+
+
+def _load_deps_of_tla_file(tla_module: Path) -> List[str]:
+    """Load all the found dependencies on disk for TLA+ module
+
+    Return:
+        a list of the contents of the dependenceis
+    """
+    print(">>>", tla_module)
+    content = tla_module.read_text()
+    deps = _get_module_deps(content)
+    print(deps)
+    return [
+        dep
+        for f in tla_module.parent.resolve().iterdir()
+        if f.stem in deps and f.suffix.lstrip(".") == TLA_SUFFIX
+        for dep in (_load_deps_of_tla_file(f) + [f.read_text()])
+    ]
+
+
+class Source:
+    """
+    A source from which the client can load data
+    """
+
+    # Supported inputs to derive a `Source`
+    # Strings should be a
+    Input = Union[str, Path]
+
+    @classmethod
+    def load_input(cls, source: Input) -> str:
+        """Convert an Input into a string:
+
+        - loading the contents of a file specified by a `Path`
+        - acting as identity on a string
+        """
+        if isinstance(source, str):
+            return source
+        elif isinstance(source, Path):
+            return source.read_text()
+        else:
+            raise ValueError(
+                "Source can only be construced from a str or a Path,"
+                f"given {type(source)}"
+            )
+
+    @classmethod
+    def of_file(cls, p: Path, aux: List[Path]) -> Self:
+        """Create a Source for use in an RPC
+
+        Args:
+            p: The main file
+            aux: Any auxiliary files required as dependencies
+        """
+        return Source(
+            source=p.read_text(),
+            aux=[s.read_text() for s in aux],
+            format=p.suffix.lstrip("."),
+        )
+
+    @classmethod
+    def of_file_load_deps(cls, p: Path) -> Self:
+        """Like `of_file` but it attempts to load dependencies from the file system
+
+        Args:
+            p: The main file
+        """
+        if not (p.suffix.lstrip(".") == TLA_SUFFIX):
+            raise ValueError(
+                f"dependencies can only be loaded for TLA files, given: {p}"
+            )
+        # The directory in which the file is located
+        source = p.read_text()
+        aux = _load_deps_of_tla_file(p)
+        return Source(source=source, aux=aux, format=TLA_SUFFIX)
+
+    def __init__(
+        self,
+        source: str,
+        *,
```

Comment:
I'm confused about the asterisk here; it's not bound to a name... is it meant to unpack arguments?
I haven't found an invocation that would use this.

Reply:
This forces all other arguments to be passed via keyword: see https://peps.python.org/pep-3102/

Reply:
I've never seen this before. Could you add a comment?

```suggestion
        *,  # force all other arguments to be keyword-only
```

Reply:
Yep good idea. Thanks.